### DISCUSSION TITLE
Is there a way to decorate the appearance of (Custom)LineBreakNode?

### DISCUSSION THREAD

**Discussion by YusukeSuzuki** [facebook/lexical#5329] [Q&A] [answered]
I am creating a CustomLineBreakNode class that inherits from LineBreakNode because I need to convert the <br> tag into another unique tag eventually. At the same time, I want to display a RETURN emoji at the line break position, so I want to make some effort in createDOM (as is well known, CSS does not provide visual decoration for the <br> tag). The code is as follows.

```JavaScript
  createDOM() {
    let fragment = document.createDocumentFragment()
    let span = document.createElement('span')
    span.innerHTML= '&#x23CE;'
    fragment.appendChild(span)
    fragment.appendChild(document.createElement('br'))
    return fragment
  }
```

If I make this work, I seem to lose control of the cursor in the text editor. The cursor is lost when I make a line break, and even if I click again to bring the cursor out, the cursor is lost again when I move across the <br> position.

Is it legal to return DocumentFragment in createDOM? Or is there a correct way to decorate <br> in Lexical?

**Comment by yf-yang** [comment]
No, createDOM returns an HTMLElement, which is not compatible with DocumentFragment
> Type '() => DocumentFragment' is not assignable to type '() => HTMLElement'.
    Type 'DocumentFragment' is missing the following properties from type 'HTMLElement': accessKey, accessKeyLabel, autocapitalize, dir, and 231 more.
    
Fortunately you can insert `<br/>` in `<span/>`
``` javascript
createDOM() {
    const span = document.createElement('span');
    span.innerHTML = '&#x23CE;';
    span.appendChild(document.createElement('br'));
    return span;
}
```

https://github.com/facebook/lexical/assets/36890796/830c2671-5ca9-482e-a3f2-d4c6cd150f31

I am not so sure because an additional issue appear: you can see the `custom-linebreak` becomes `linebreak` again after typing one more character, the node replacement does not work any more. Maybe that's another bug.

**Reply by yf-yang** [reply]
After debugging the code, I find the node get replaced to linebreak because I do not implement the clone method. However, the delete key still not work because the dom selection cannot correctly map to the `CustomLineBreakNode` here.

I can come up with an alternative with pseudo CSS.
``` javascript
class CustomLineBreakNode extends LineBreakNode {
  static getType(): string {
    return 'custom-linebreak';
  }

  static clone(node) {
    return new CustomLineBreakNode(node.__key);
  }

  createDOM() {
    const dom = document.createElement('br');
    dom.className = 'custom-linebreak';
    return dom;
  }
}
```
However, it seems there is some issue with `<br/>` tag and the psuedo CSS. Check [this SO question](https://stackoverflow.com/questions/1672879/can-you-add-line-breaks-to-the-after-pseudo-element) for more details. If you find a way out, I am also interested.

**Reply by yf-yang** [reply]
Well, I don't think it is a good idea to custom br in content editable. The closest solution is https://stackoverflow.com/a/28342884, but in the content editable, the anchor position is still wrong. Not sure if you can find a better solution.

``` css
.custom-linebreak {
  content: ''
}

.custom-linebreak:before {
  content: '↩\A'
}
```

**Reply by YusukeSuzuki** [reply]
Thanks for the analysis. But it seems that the approach described in the SO often doesn't work well with modern browsers.

Your use of span gave me an idea of how to do it without the br tag, and I think that's how it would work in HTML. I think it is possible to use exportDOM to replace the span tag with a br tag.

https://jsfiddle.net/uy30x2tg/1/

```css
:has(+ span.pseudo-br):after {
  content: '\23ce';
}

span.pseudo-br {
  display: block;
  clear: inline-start;
}
```

```html
<div>
  <span>this is text</span><span class="pseudo-br"></span>
  <span class="pseudo-br"></span>
  <span>this is text</span><span class="pseudo-br"></span>
  <span class="pseudo-br"></span>
  <span class="pseudo-br"></span>
  <span>this is text</span><span class="pseudo-br"></span>
  <span>this is text</span><span class="pseudo-br"></span>
  <span>this is text</span>
</div> 
```

I was concerned about the strange cursor behavior when LineBreakNode is inherited, but I thought it was a DocumentFragment issue. I will try to create a CustomLineBreakNode using the HTML/CSS approach above, but if the behavior is still strange, I will try to use DecoratorNode as a base. Then rewrite the key events to insert a Decorator instead of a LineBreak.

**Reply by YusukeSuzuki** [reply, marked as answer]
Okay. Finally, by returning an empty \<span\> in CustomLineBreakNode's createDOM and applying the CSS I wrote above, I was able to display the RETURN mark while simulating the behavior of the \<br\> tag. Thanks a lot for your advice.
